File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        
        stack<char> stack;
        
        for (char c : s) {
            if (c == '(' || c == '{' || c == '[') {
                stack.push(c);
            }
            else if (!stack.empty() && c == ')' ) {
                if (stack.top() == '(') {
                    stack.pop();
                }
                else {
                    return false;
                }
            }
            else if (!stack.empty() && c == '}' ) {
                if (stack.top() == '{') {
                    stack.pop();
                }
                else {
                    return false;
                }
            }
            else if (!stack.empty() && c == ']' ) {
                if (stack.top() == '[') {
                    stack.pop();
                }
                else {
                    return false;
                }
            }
            else {
                return false;
            }
        }
        
        if (stack.empty()) return true;
        
        return false;
        
    }
};

File: 0020-valid-parentheses/0020-valid-parentheses.scala
import scala.collection.mutable
import scala.util.boundary, boundary.break

object Solution {
    def isValid(s: String): Boolean = {
        boundary {
            val stack = mutable.Stack[Char]()
            val pairs = Map(')' -> '(', '}' -> '{', ']' -> '[')

            for (char <- s) {
                if (pairs.contains(char)) {
                    if (stack.isEmpty || stack.pop() != pairs(char)) break(false)
                }
                else {
                    stack.push(char)
                }
            }
            stack.isEmpty
        }


        
    }
}